### PR TITLE
Merge staging into development

### LINKED ISSUE #15: Current DAG structure critical issue
Currently a block is added to both the receiver and sender's blockchain. This works provided the sender and the receiver have exactly the same block height. If they do not then there will be two blocks at certain heights. To solve this a block can be one of two types:
* Send 
* Receive
Send blocks are made by the sender and are put straight into the mempool after being validated. Once they have at least 2/3 signatures from the committee then the block is sent to the committee dealing with the receiver. The receiving committee then forms a receive block with the exact same data as the send block. Unlike send blocks, receiver blocks do not have to be signed by the wallet. The are created by the committee that deals with that chain and to be valid must reference a block that is in the mempool. If a node receives a receive block it first looks in the mempool for the block that the receive block references. If it does not have it then it asks the node it got the receiving block from to sen the corresponding send block. If the send block is valid and all fields of the receive block other than hash, timestamp, height, type or signatures are the same and the node signatures are valid then it accepts both blocks. If a node does not respond with a send block related to the receive block then the node simply ignores the block. 
[POLL] 
Should we add the receive block without a parent (the send block) to the mempool, tell me what you think on this one as i cant decide.

Some edits that need to be done:

-  Never enact a send block, you only enact receive blocks
- Do not save a send block until we see a (valid) receive block, until then add it to the mempool and wait
- Periodically ask peers for a receive block relating to the send block, if after 3 hours we have not got one purge the block from the mempool until we next see it.
- Add the following parameters to block:
 - block_type: Custom enum?
 - send_block_hash: Option<String>
 - anything else?

Im going to make a separate branch and work on this, if anyone want to help out that would be great. Ill be adding code documentation as i go along as well as fixing any warnings

### PR DESCRIPTION
**Pull request by leahcornelius**
This merges both fix-the-dag and p2p-refactor into development (#20 #15, #14)
- [x] Fix all compiler warnings caused by the new functions (99% naming issues)
- [ ] Write unit tests for new code

### CODE REVIEW

**Comment by leahcornelius** [comment]
@TheDevMinerTV if you get a chance can you review just the p2p refactor aspect and ensure the are no concerns on your end?

**Comment by leahcornelius** [comment]
Feel free to take a look at the dag to restructure one and see if anything looks blatantly wrong, but bare in mind its not done yet so there are some bits that are off.

File: p2p/src/core.rs
Comment (TheDevMinerTV) [inline p2p/src/core.rs]: misspelled

File: p2p/src/core.rs
Comment (TheDevMinerTV) [inline p2p/src/core.rs:156]: These should be implemented, no?

File: p2p/src/handle.rs
Comment (TheDevMinerTV) [inline p2p/src/handle.rs]: This should be configurable

File: p2p/src/helper.rs
Comment (TheDevMinerTV) [inline p2p/src/helper.rs]: This should probably do something.

File: p2p/src/message_types.rs
Comment (TheDevMinerTV) [inline p2p/src/message_types.rs]: I found both in the old code, you're free to remove one :)

File: p2p/src/server.rs
Comment (TheDevMinerTV) [inline p2p/src/server.rs]: Should be `decrement_connections`.

**Review by TheDevMinerTV — DISMISSED** [review]

**Comment by leahcornelius** [comment]
Awesome thanks ill fix those in a bit, also reminded me of a couple of things p2p wise I need to implement regarding sharing your peerlist

**Comment by richreggio** [comment]
So I gave trying to merge the staging branch with the development branch a go and also worked on removing any new compiler warnings that weren't talking about unused code or imports. Also went through and changed most of what clippy would complain about, mainly explicit returns at the end of functions and using &String as parameters when &str would suffice. There were a couple of instances of integral math that always returned 0 but that is obviously not what was intended so i switched it to floating point (was a part of the parallelism code). Also implemented the Display trait in place of a couple of to_string() functions since doing so implements to_string(). I'm still wrapping my head around parts of the code base and learning the finer points of Rust but I really am impressed by how much work must have gone into this.

This should finish? PR #26 

Not really sure, I've only ever worked on personal repo's and these are my first attempts at contributing to an open source project.

I'm not sure how I could add to this PR or if I have to open a new one.

[Fork with changes](https://github.com/richreggio/avrio-rs)

**Comment by leahcornelius** [comment]
That's absolutely incredible thanks! I haven't been able to work on avrio for the last couple of days but tomorrow i will take a look at all your changes and fix anything i can see. The code base is currently quite chaotic as unfortunately i made a naive mistake of "i don't need to document this, i'll remember it". Luckily i can read my code and understand what it does, but i am going through and writing comments for each bit as i go along. That being said, if you do need me to explain any part of the code or protocol do just ask. I really appreciate the contributions, and if you are interested in helping out more i'm more than happy to answer any questions!

**Comment by leahcornelius** [comment]
merged your changes int staging, will review and change anything needed. I need to do some small changes as pointed out by dev miner before merging this PR, and i want to be 100% sure that the cross chain transactions issue was not introduced by these changes so i will merge this once that issue is fixed. After that the rest of the stuff needing to be done for a stable release (eg double checking block hashes with multiple peers, gracefully handling wrong messages during sync, not sending peers who are syncing messages etc) as well as finishing off the API and moving the wallet external can be done on development

**Comment by leahcornelius** [comment]
Conflicts? How odd...

File: p2p/src/helper.rs
Comment (leahcornelius) [reply p2p/src/helper.rs]: No, i can remove that code. Functions should not unlock the passed peers as that is up to the parent thread.

File: p2p/src/server.rs
Comment (leahcornelius) [reply p2p/src/server.rs]: Solved, also i need to call those functions

File: p2p/src/message_types.rs
Comment (leahcornelius) [reply p2p/src/message_types.rs]: Going to leave for now

**Pull request merged**
